Confusion between syntax for pattern alternative and typed pattern on a union type

As in the title, running this code in REPL throws `java.lang.VerifyError`:

```
scala> def f(): String | Int = 0
def f(): String | Int
scala> f() match { case s: String | Int => s }
```

```
java.lang.VerifyError: Bad local variable type
Exception Details:
  Location:
    rs$line$2$.<init>()V @33: aload_3
  Reason:
    Type top (current frame, locals[3]) is not assignable to reference type
  Current Frame:
    bci: @33
    flags: { }
    locals: { 'rs$line$2$', 'java/lang/Object' }
    stack: { 'rs$line$2$' }
  Bytecode:
    0x0000000: 2ab7 0014 2ab3 0016 2ab2 001b b600 1f4c
    0x0000010: 2bc1 0021 9900 172b c000 214d 2c4e a700
    0x0000020: 072d a700 06a7 fffc a700 1cb2 0026 2bb6
    0x0000030: 002a 9900 09a7 ffec 0000 bfbb 002c 592b
    0x0000040: b700 2fbf b500 31b1
  Stackmap Table:
    full_frame(@33,{Object[#2],Object[#4]},{Object[#2]})
    full_frame(@37,{Object[#2],Object[#4],Object[#33],Object[#33]},{Object[#2]})
    full_frame(@40,{Object[#2],Object[#4]},{Object[#2],Top})
    same_locals_1_stack_item_frame(@43,Object[#2])
    full_frame(@56,{},{Object[#53]})
    full_frame(@59,{Object[#2],Object[#4]},{Object[#2]})
    full_frame(@68,{Object[#2],Object[#4]},{Object[#2],Top})

	at rs$line$2.res0Show(rs$line$2)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at dotty.tools.repl.Rendering.valueOf(Rendering.scala:58)
	at dotty.tools.repl.Rendering.renderVal(Rendering.scala:79)
	at dotty.tools.repl.ReplDriver.$anonfun$displayDefinitions$8(ReplDriver.scala:269)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at scala.collection.TraversableLike.map(TraversableLike.scala:234)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at dotty.tools.repl.ReplDriver.displayMembers$1(ReplDriver.scala:269)
	at dotty.tools.repl.ReplDriver.$anonfun$displayDefinitions$15(ReplDriver.scala:296)
	at scala.Option.map(Option.scala:146)
	at dotty.tools.repl.ReplDriver.$anonfun$displayDefinitions$13(ReplDriver.scala:294)
	at dotty.tools.dotc.core.Periods.atPhase(Periods.scala:26)
	at dotty.tools.dotc.core.Phases.atPhase(Phases.scala:36)
	at dotty.tools.dotc.core.Phases.atPhase$(Phases.scala:35)
	at dotty.tools.dotc.core.Contexts$Context.atPhase(Contexts.scala:57)
	at dotty.tools.repl.ReplDriver.displayDefinitions(ReplDriver.scala:289)
	at dotty.tools.repl.ReplDriver.$anonfun$compile$2(ReplDriver.scala:226)
	at scala.util.Either.fold(Either.scala:188)
	at dotty.tools.repl.ReplDriver.compile(ReplDriver.scala:220)
	at dotty.tools.repl.ReplDriver.interpret(ReplDriver.scala:190)
	at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:143)
	at dotty.tools.repl.Main$.main(Main.scala:6)
	at dotty.tools.repl.Main.main(Main.scala)
```

Parenthesising the type, however, runs normally:
```
scala> f() match { case s: (String | Int) => s }
val res2: Any = 0
```

This code should not compile, as in scalac where you get:
```scala
try/i3332.scala:3: error: illegal variable in pattern alternative
    (1: Any) match { case (s: String) | _: Int => s }
                           ^
```
There's already an issue for that: https://github.com/lampepfl/dotty/issues/1612

The other issue is how to parse `case s: String | Int`, it's unfortunate that the current meaning is confusing when combined with union types, but I don't know if there's a good solution here

--- a/compiler/src/dotty/tools/dotc/core/Mode.scala
+++ b/compiler/src/dotty/tools/dotc/core/Mode.scala
@@ -8,7 +8,7 @@ case class Mode(val bits: Int) extends AnyVal {
   def &~ (that: Mode) = Mode(bits & ~that.bits)
   def is (that: Mode) = (bits & that.bits) == that.bits
 
-  def isExpr = (this & PatternOrType) == None
+  def isExpr = (this & PatternOrTypeBits) == None
 
   override def toString =
     (0 until 31).filter(i => (bits & (1 << i)) != 0).map(modeName).mkString("Mode(", ",", ")")
@@ -41,6 +41,9 @@ object Mode {
 
   /** We are looking at the arguments of a supercall */
   val InSuperCall = newMode(6, "InSuperCall")
+
+  /** We are in a pattern alternative */
+  val InPatternAlternative = newMode(7, "InPatternAlternative")
 
   /** Allow GADTFlexType labelled types to have their bounds adjusted */
   val GADTflexible = newMode(8, "GADTflexible")
@@ -87,7 +90,7 @@ object Mode {
   /** Don't suppress exceptions thrown during show */
   val PrintShowExceptions = newMode(18, "PrintShowExceptions")
 
-  val PatternOrType = Pattern | Type
+  val PatternOrTypeBits = Pattern | Type | InPatternAlternative
 
   /** We are elaborating the fully qualified name of a package clause.
    *  In this case, identifiers should never be imported.

--- a/compiler/src/dotty/tools/dotc/typer/Namer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Namer.scala
@@ -936,10 +936,10 @@ class Namer { typer: Typer =>
   }
 
   def typedAheadType(tree: Tree, pt: Type = WildcardType)(implicit ctx: Context): tpd.Tree =
-    typedAheadImpl(tree, typer.typed(_, pt)(ctx retractMode Mode.PatternOrType addMode Mode.Type))
+    typedAheadImpl(tree, typer.typed(_, pt)(ctx retractMode Mode.PatternOrTypeBits addMode Mode.Type))
 
   def typedAheadExpr(tree: Tree, pt: Type = WildcardType)(implicit ctx: Context): tpd.Tree =
-    typedAheadImpl(tree, typer.typed(_, pt)(ctx retractMode Mode.PatternOrType))
+    typedAheadImpl(tree, typer.typed(_, pt)(ctx retractMode Mode.PatternOrTypeBits))
 
   def typedAheadAnnotation(tree: Tree)(implicit ctx: Context): Symbol = tree match {
     case Apply(fn, _) => typedAheadAnnotation(fn)

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -1211,12 +1211,14 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
             typed(untpd.Bind(tree.name, untpd.TypedSplice(arg)).withPos(tree.pos), arg.tpe) :: Nil)
       case _ =>
         val sym = newPatternBoundSym(tree.name, body1.tpe, tree.pos)
+        if (ctx.mode.is(Mode.InPatternAlternative)) ctx.error("Illegal variable in pattern alternative", tree.pos)
         assignType(cpy.Bind(tree)(tree.name, body1), sym)
     }
   }
 
   def typedAlternative(tree: untpd.Alternative, pt: Type)(implicit ctx: Context): Alternative = track("typedAlternative") {
-    val trees1 = tree.trees mapconserve (typed(_, pt))
+    val nestedCtx = ctx.addMode(Mode.InPatternAlternative)
+    val trees1 = tree.trees.mapconserve(typed(_, pt)(nestedCtx))
     assignType(cpy.Alternative(tree)(trees1), trees1)
   }
 
@@ -1754,7 +1756,7 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
         Inliner.removeInlineAccessors(mdef.symbol)
 
   def typedExpr(tree: untpd.Tree, pt: Type = WildcardType)(implicit ctx: Context): Tree =
-    typed(tree, pt)(ctx retractMode Mode.PatternOrType)
+    typed(tree, pt)(ctx retractMode Mode.PatternOrTypeBits)
   def typedType(tree: untpd.Tree, pt: Type = WildcardType)(implicit ctx: Context): Tree = // todo: retract mode between Type and Pattern?
     typed(tree, pt)(ctx addMode Mode.Type)
   def typedPattern(tree: untpd.Tree, selType: Type = WildcardType)(implicit ctx: Context): Tree =

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -1210,9 +1210,13 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
         tpd.cpy.UnApply(body1)(fn, Nil,
             typed(untpd.Bind(tree.name, untpd.TypedSplice(arg)).withPos(tree.pos), arg.tpe) :: Nil)
       case _ =>
-        val sym = newPatternBoundSym(tree.name, body1.tpe, tree.pos)
-        if (ctx.mode.is(Mode.InPatternAlternative)) ctx.error("Illegal variable in pattern alternative", tree.pos)
-        assignType(cpy.Bind(tree)(tree.name, body1), sym)
+        if (tree.name == nme.WILDCARD) body1
+        else {
+          val sym = newPatternBoundSym(tree.name, body1.tpe, tree.pos)
+          if (ctx.mode.is(Mode.InPatternAlternative))
+            ctx.error(i"Illegal variable ${sym.name} in pattern alternative", tree.pos)
+          assignType(cpy.Bind(tree)(tree.name, body1), sym)
+        }
     }
   }
 

--- a/tests/neg/i3332.scala
+++ b/tests/neg/i3332.scala
@@ -1,0 +1,9 @@
+object Main {
+  def main(args: Array[String]): Unit = {
+    (1: Any) match {
+      case x: String | Int => "OK"
+      case (_: String) | (_: Int) => "OK"
+      case (s: String) | _: Int => s   // error: Illegal variable in pattern alternative
+    }
+  }
+}

--- a/tests/neg/i3332.scala
+++ b/tests/neg/i3332.scala
@@ -1,7 +1,7 @@
 object Main {
   def main(args: Array[String]): Unit = {
     (1: Any) match {
-      case x: String | Int => "OK"
+      case x: String | Int => "OK"  // error: Illegal variable in pattern alternative
       case (_: String) | (_: Int) => "OK"
       case (s: String) | _: Int => s   // error: Illegal variable in pattern alternative
     }

--- a/tests/neg/i3332.scala
+++ b/tests/neg/i3332.scala
@@ -7,3 +7,11 @@ object Main {
     }
   }
 }
+
+// #1612
+object Test {
+  def g(p:(Int,Int)) = p match {
+    case (10,n) | (n,10) => println(n) // error // error (Illegal variable in pattern alternative)
+    case _ => println("nope")
+  }
+}

--- a/tests/neg/i3332.scala
+++ b/tests/neg/i3332.scala
@@ -10,8 +10,12 @@ object Main {
 
 // #1612
 object Test {
+  case class A()
   def g(p:(Int,Int)) = p match {
     case (10,n) | (n,10) => println(n) // error // error (Illegal variable in pattern alternative)
     case _ => println("nope")
+  }
+  def test(x: Any) = x match {
+    case _: String | _ @ A() => 1
   }
 }

